Mdsclient

mfromano opened the pull request:
I added 3 new urls to urls.py, each of which serves as a REST endpoint for a client who wishes to retrieve notifications based on patient ID, procedure ID,  or who wish to retrieve the entire list of notifications. In addition, I added the corresponding "get" methods to json.py. 

Edit: just fixed a fairly obvious bug: the case where multiple notifications exist for a given patient or procedure id. Still not sure what/if there is a unique identifier for notifications, though I think that adding a field like "date" would be helpful.

bjacobel commented:
:thumbsup: will take a look at this Friday or Saturday!

mfromano commented:
Hey, updated my pull request. I just fixed a couple of obvious bugs (the cases where multiple notifications are returned for a given procedure_id or patient_id). I'm still unsure as to what/if there is a unique identifier for a given notification. I think that adding a "date and time" field would be a good idea, which is what I'm going to do next if you agree. This could serve as a sort of unique identifier and would at least help to filter based on most recent.

bjacobel commented on sana/mrs/json.py:
```diff
@@ -1189,5 +1191,151 @@ def syc_encounters(request, patient_id, encounters=None):
+            notification = Notification.objects.filter(patient_id=id)
+            logging.info("we finished getting the notification")
+            response = {'status': 'SUCCESS',
+                    'data': [cjson.decode(d.to_json()) for d in notification],
```
So there's three cases here:

| result of `Notification.objects.filter(patient_id=id)` | `response['data']` |
| --- | --- |
| exactly one object | `<that object>` |
| exactly 0 objects | `"Problem while getting notification: <error>"` |
| > 1 object | `[<object 1>, <object 2>, <etc>]` |

That puts three different datatypes into `response['data']` in each situation: in the first case, it's a JSON object, in the second it's a string and in the third it's an array. 

The way I would handle this is to instead always return an array in `response['data']`. In the first case, it can be an array with just one element. In the second case, it can be an array with no elements (you can pass the status string along in another field of `response`, like `response['message']`. Then in the third case, do as you're already doing. The advantage of that is that the client gets one consistent datatype in the `data` field - it doesn't need any conditional logic around what data structure to deserialize this to if it knows it's always going to be an array.

If you do this, you can condense this `try`/`except`/`except` into just one call to `Notification.objects.filter(patient_id=id)`. That will return the array you want to serialize, be it empty or having one object or multiple or whatever.

That also lets you get away from this structure where you have a try/except inside of a try/except which can be a little difficult to follow along with.

Hope that makes sense!

bjacobel commented on sana/mrs/json.py:
```diff
@@ -1189,5 +1191,151 @@ def syc_encounters(request, patient_id, encounters=None):
+            'status': 'FAILURE',
+            'data': "Problem while getting notification list: %s" % e,
+        }
+    return HttpResponse(cjson.encode(response), mimetype=("application/json; charset=utf-8"))
```
Nothing to do about this while we're on Django 1.3, but worth noting that Django 1.5 [deprecates the `mimetype` argument in the `HttpResponse` constructor](https://docs.djangoproject.com/en/1.5/ref/request-response/#id5) in favor of `content_type` and Django 1.6 removes it entirely. Will need to track all these down and replace then when/if we upgrade Djangos.

bjacobel commented on sana/mrs/views.py:
```diff
@@ -94,7 +94,7 @@ def procedure_submit(request):
 
 def list_notifications(request):
-    """Queries the db for all notifiactions"""
+    """Queries the db for all notifications"""
```
lol, good catch

bjacobel commented:
@mfromano This looks great. Re: unique IDs for notifications: Django should be storing each with a `pk` even if you don't specify it on the model. I believe, so that's a way of getting a unique ID in the MDS at least (can't correlate that back to an ID in the MRS, unfortunately).

If you do add a datetime, I would look into something like:

`created_at = models.DateTimeField(auto_now_add=True)` -- that will set the time to whenever the model instance was created automatically. Docs on it [here](https://django.readthedocs.org/en/1.3.X/ref/models/fields.html#django.db.models.DateTimeField).

The tricky part is going to be migrations - this is Django 1.3 so it doesn't have any migration framework built in. You'll have to look into [South](http://south.aeracode.org/) for migration support, unfortunately. And it doesn't look like there are any migrations already in the project so you won't have any examples here to go off of. Let me know if you decide to dive into this and I'll see what I can do to help.

mfromano commented:
Hi Brian,

I added South for migrations, added a requirements.txt, made the changes to the REST responses, and updated the Notification model to include a DateTimeField. Let me know what you think.

Best,
Mike

bjacobel commented:
Couple things:
- File mode changed on a lot of files - not sure what's up with that
- I set up a [requirements manifest structure](https://github.com/addisclinic/mobile-dispatch-server/tree/develop/reqs) that's been merged to develop already - drop South into one of those files, probably `base.txt` because it'll need to be used in both prod and dev
- I think that's why the Travis build failed -- it couldn't find South in those files where it was expecting it

mfromano commented:
Okay, I made those changes and it appeared to work. Any idea how to revert the file permissions other than going one by one? I accidentally recursively "chmod +777"'d the whole directory...

mfromano commented:
Nevermind, I finished fixing all of the file permissions. We should be good.
